### ISSUE TITLE
BUG: Operator equals works where np.equal does not

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#9147]
```python
>>> x = np.array([(0, 1, 0)] * 10, dtype=[('f0', '<i4'), ('f1', '<i4'), ('f2', '<i4')])
>>> x == x
array([ True,  True,  True,  True,  True,  True,  True,  True,  True,  True], dtype=bool)
>>> np.equal(x, x)
FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
NotImplemented
```

**Comment by shoyer** [comment]
I think this is an unintended and/or unforeseen consequence of a partial fix for issues with `==` in old versions of numpy (see #5964).

As you can see, this deprecation notice went in about two years ago now (NumPy 1.10). Can we finally fix the broken behavior? It seems like we've given more than enough warning to anyone who may have been relying on the old broken behavior.

**Comment by charris** [comment]
@shoyer The 1.10 deprecations/futurwarnings will expire in 1.14. There are several that I will be putting in a PR soonish to change.

**Comment by seberg** [comment]
ufuncs currently don't support flexible types, so string and voids are specifically handled by `__eq__` and not send to `np.equal`.

**Comment by mhvk** [comment]
The special handling gets the elements and compares them in turn -- and those comparisons do get send to any `__array_ufunc__`, as does the combination of them via `logical_and`, so from that perspective it is OK at least for `ndarray` subclasses.

For array-likes it is perhaps not unreasonable for the mixin to not try to also support structured array-likes in a similar way. Though it would be good to document this.

**Comment by mhvk** [comment]
Looking in more detail at the source for `array_richcompare` at `Py_EQ` ( https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/arrayobject.c#L1302), I see that the `RICHCMP_GIVE_UP_IF_NEEDED` comes after the check for `void`, rather than before, which seems wrong. See new issue #9153.

Similarly, the check for string comes even before any of this, which may also not be quite right. Looking in detail, it falls through if `other` is not also a string-like array, so this case works for array-likes. And for subclasses, since the `ufunc` doesn't work for strings, I guess one has little choice but to do what is done there; subclasses can override simply by defining `__eq__` and their `__eq__` will get called.

**Comment by ahaldane** [comment]
Just want to say that I just hit this bug while trying to implement `np.equal` in `__array_ufunc__` for my Masked-ducktype.  More in detail, my ducktype inherits from `numpy.lib.mixins.NDArrayOperatorsMixin` which implements `    __eq__ = _binary_method(um.equal, 'eq')`. In other words it assumes `__eq__` can be replaced by `np.equal`, however this bug means that is not true for structured arrays.

**Comment by mhvk** [comment]
Yes, structured arrays are a bit of a step-child currently, with the difficulty of defining ufuncs for them

**Comment by seberg** [comment]
Changed the title, strings are in the ufunc now.  Void could be moved, although it is a bit of a smaller project to do so.
A new difference is that `==` will return an array of all `False` when dtypes cannot be compared while `np.equal()` errors.

**Comment by mattip** [comment]
Closing this as stale.
